File: backend/Models/Items/items/item.py
# coding=utf-8

from datetime import datetime
from sqlalchemy import Column, String, Boolean, Integer, Float, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import ARRAY
from Models.entity import Entity
from Configuration.config import DB_SESSION, ENGINE, BASE
from Models.sqlal_mutable_array import MutableList
from marshmallow import Schema, fields

class ItemTable(Entity, BASE):
    __tablename__ = 'items'

    id = Column(Integer, primary_key=True)
    
    name = Column(String)
    photo = Column(String)
    city = Column(String)
    state = Column(String)
    zipcode = Column(Float)
    description = Column(String)
    instructions = Column(String)
    claimed_status = Column(Boolean)
    taken_status = Column(Boolean)
    owner_id = Column(Integer)
    sponsored = Column(Boolean)
    selected_user_id = Column(Integer)
    approved = Column(Boolean)
    claimed_at = Column(DateTime)

    search = Column(String)
    favorites = Column(Integer)
    claims = Column(Integer)
    
    def __init__(self, name, photo, city, state, zipcode, description, owner_id, created_by):
        Entity.__init__(self, created_by)
        self.name = name
        self.photo = photo
        self.city = city
        self.state = state
        self.zipcode = zipcode
        self.description = description
        self.owner_id = owner_id
        self.claimed_status = False
        self.taken_status = False
        self.approved = True
        self.favorites = 0
        self.claims = 0

class ItemTableSchema(Schema):
    id = fields.Number()

    name = fields.Str()
    photo = fields.Str()
    city = fields.Str()
    state = fields.Str()
    zipcode = fields.Number()
    description = fields.Str()
    instructions = fields.Str()
    claimed_status = fields.Boolean()
    taken_status = fields.Boolean()
    owner_id = fields.Number()
    selected_user_id = fields.Number()
    sponsored = fields.Boolean()
    approved = fields.Boolean()
    claimed_at = fields.DateTime()

    favorites = fields.Number()
    claims = fields.Number()

    created_at = fields.DateTime()
    updated_at = fields.DateTime()
    last_updated_by = fields.Str()

File: frontend/src/Helpers/Items/Classes/Item.ts
import { IItem } from '../Interfaces/IItem';

export class Item implements IItem {
    private id: number;
    private name: string;
    private photo: string;
    private description: string;
    private takenStatus: boolean;
    private claimedStatus: boolean;
    private sponsored: boolean;
    private ownerId: number;
    private claimedUserId: number;
    private claimedAt: Date;
    private createdAt: Date;
    private city: string;
    private state: string;
    private zipcode: number;
    private instructions: string;
    private favorites: number;
    private claims: number;

    public get Instructions(): string {
        return this.instructions;
    }
    public set Instructions (value: string) {
        this.instructions = value;
    }

    public get Id(): number {
        return this.id;
    }
    public set Id (value: number) {
        this.id = value;
    }

    public get Name(): string {
        return this.name;
    }
    public set Name (value: string) {
        this.name = value;
    }

    public get Photo(): string {
        return this.photo;
    }
    public set Photo (value: string) {
        this.photo = value;
    }

    public get Description(): string {
        return this.description;
    }
    public set Description (value: string) {
        this.description = value;
    }
    public get ClaimedStatus(): boolean {
        return this.claimedStatus;
    }
    public set ClaimedStatus (value: boolean) {
        this.claimedStatus = value;
    }
    public get TakenStatus(): boolean {
        return this.takenStatus;
    }
    public set TakenStatus (value: boolean) {
        this.takenStatus = value;
    }
    public get Sponsored(): boolean {
        return this.sponsored;
    }
    public set Sponsored (value: boolean) {
        this.sponsored = value;
    }
    public get OwnerId(): number {
        return this.ownerId;
    }
    public set OwnerId (value: number) {
        this.ownerId = value;
    }
    public get ClaimedUserId(): number {
        return this.claimedUserId;
    }
    public set ClaimedUserId (value: number) {
        this.claimedUserId = value;
    }
    public get City(): string {
        return this.city;
    }
    public set City (value: string) {
        this.city = value;
    }
    public get State(): string {
        return this.state;
    }
    public set State (value: string) {
        this.state = value;
    }
    public get Zipcode(): number {
        return this.zipcode;
    }
    public set Zipcode (value: number) {
        this.zipcode = value;
    }

    public get Favorites(): number {
        return this.favorites;
    }
    public set Favorites (value: number) {
        this.favorites = value;
    }
    public get Claims(): number {
        return this.claims;
    }
    public set Claims (value: number) {
        this.claims = value;
    }

    public get CreatedAt(): Date {
        return this.createdAt;
    }
    public set CreatedAt (value: Date) {
        this.createdAt = value;
    }
    public get ClaimedAt(): Date {
        return this.claimedAt;
    }
    public set ClaimedAt (value: Date) {
        this.claimedAt = value;
    }

    constructor(id: number, name:string, photo:string, description:string, ownerId:number, city:string, state:string, zipcode: number, takenStatus: boolean, claimedStatus: boolean, sponsored: boolean, claimedUserId: number, createdAt: Date, claimedAt: Date) {
        this.id = id;
        this.name = name;
        this.photo = photo;
        this.description = description;
        this.ownerId = ownerId;
        this.city = city;
        this.state = state;
        this.zipcode = zipcode;
        this.sponsored = sponsored;
        this.takenStatus = takenStatus;
        this.claimedStatus = claimedStatus;
        this.claimedUserId = claimedUserId;
        this.createdAt = createdAt;
        this.claimedAt = claimedAt;
        this.instructions = '';
        this.favorites = 0;
        this.claims = 0;
    }
    
}